obsreport/obsreporttest/otelprometheuschecker.go

```diff
@@ -0,0 +1,138 @@
+
+	io_prometheus_client "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/expfmt"
+	"go.opencensus.io/stats/view"
+	"go.opentelemetry.io/otel/attribute"
+	"go.uber.org/multierr"
+
+	"go.opentelemetry.io/collector/config"
+)
+
+// prometheusChecker is used to assert exported metrics from a prometheus handler.
+type prometheusChecker struct {
+	promHandler http.Handler
+}
+
+func (pc *prometheusChecker) checkReceiverTraces(receiver config.ComponentID, protocol string, acceptedSpans, droppedSpans int64) error {
+	receiverAttrs := attributesForReceiverMetrics(receiver, protocol)
+	return multierr.Combine(
+		pc.checkCounter("receiver_accepted_spans", acceptedSpans, receiverAttrs),
+		pc.checkCounter("receiver_refused_spans", droppedSpans, receiverAttrs))
+}
+
+func (pc *prometheusChecker) checkReceiverLogs(receiver config.ComponentID, protocol string, acceptedLogRecords, droppedLogRecords int64) error {
+	receiverAttrs := attributesForReceiverMetrics(receiver, protocol)
+	return multierr.Combine(
+		pc.checkCounter("receiver_accepted_log_records", acceptedLogRecords, receiverAttrs),
+		pc.checkCounter("receiver_refused_log_records", droppedLogRecords, receiverAttrs))
+}
+
+func (pc *prometheusChecker) checkReceiverMetrics(receiver config.ComponentID, protocol string, acceptedMetricPoints, droppedMetricPoints int64) error {
+	receiverAttrs := attributesForReceiverMetrics(receiver, protocol)
+	return multierr.Combine(
+		pc.checkCounter("receiver_accepted_metric_points", acceptedMetricPoints, receiverAttrs),
+		pc.checkCounter("receiver_refused_metric_points", droppedMetricPoints, receiverAttrs))
+}
+
+func (pc *prometheusChecker) checkCounter(expectedMetric string, value int64, attrs []attribute.KeyValue) error {
+	// Forces a flush for the opencensus view data.
+	_, _ = view.RetrieveData(expectedMetric)
+
+	ts, err := pc.getMetric(expectedMetric, io_prometheus_client.MetricType_COUNTER, attrs)
+	if err != nil {
+		return err
+	}
+
+	expected := float64(value)
+	if math.Abs(expected-ts.GetCounter().GetValue()) > 0.0001 {
+		return fmt.Errorf("values for metric '%s' did no match, expected '%f' got '%f'", expectedMetric, expected, ts.GetCounter().GetValue())
+	}
+
+	return nil
+}
+
+// getMetric returns the metric time series that matches the given name, type and set of attributes
+// it fetches data from the prometheus endpoint and parse them, ideally OTel Go should provide a MeterRecorder of some kind.
+func (pc *prometheusChecker) getMetric(expectedName string, expectedType io_prometheus_client.MetricType, expectedAttrs []attribute.KeyValue) (*io_prometheus_client.Metric, error) {
+	parsed, err := fetchPrometheusMetrics(pc.promHandler)
+	if err != nil {
+		return nil, err
+	}
+
+	metricFamily, ok := parsed[expectedName]
+	if !ok {
+		// OTel Go adds `_total` suffix for all monotonic sum.
+		// TODO: once remove support for OpenCensus, we should always check for metrics with the `_total` suffix for prometheus counter metrics.
```

Comment:
This feels like cheating but I couldn't find a better way, we can't disable this behavior even for tests.

Reply:
Once we remove support for OC, we should use a "TestExporter" for otel metrics and not check prometheus :))

Reply:
Sure! I'll just remove this TODO, since it doesn't really makes sense :)